Firestore/core/src/firebase/firestore/core/sync_engine.mm

```diff
@@ -0,0 +1,518 @@
+  auto it = limbo_resolutions_by_target_.find(target_id);
+  if (it != limbo_resolutions_by_target_.end() &&
+      it->second.document_received) {
+    return DocumentKeySet{it->second.key};
+  } else {
+    auto found = query_views_by_target_.find(target_id);
+    if (found != query_views_by_target_.end()) {
+      return found->second->view().synced_documents();
+    }
+    return DocumentKeySet{};
+  }
+}
+
+void SyncEngine::NotifyUser(BatchId batch_id, Status status) {
+  auto it = mutation_callbacks_.find(current_user_);
+
+  // NOTE: Mutations restored from persistence won't have callbacks, so
+  // it's okay for this (or the callback below) to not exist.
+  if (it == mutation_callbacks_.end()) {
+    return;
+  }
+
+  std::unordered_map<BatchId, StatusCallback>& callbacks = it->second;
+  auto callback_it = callbacks.find(batch_id);
+  if (callback_it != callbacks.end()) {
+    callback_it->second(std::move(status));
+    callbacks.erase(callback_it);
+  }
+}
+
+void SyncEngine::TriggerPendingWriteCallbacks(BatchId batch_id) {
+  auto it = pending_writes_callbacks_.find(batch_id);
+  if (it != pending_writes_callbacks_.end()) {
+    for (const auto& callback : it->second) {
+      callback(Status::OK());
+    }
+
+    pending_writes_callbacks_.erase(it);
+  }
+}
+
+void SyncEngine::FailOutstandingPendingWriteCallbacks(
+    absl::string_view message) {
+  for (const auto& entry : pending_writes_callbacks_) {
+    for (const auto& callback : entry.second) {
+      callback(Status(Error::Cancelled, message));
+    }
+  }
+
+  pending_writes_callbacks_.clear();
+}
+
+void SyncEngine::EmitNewSnapshotsAndNotifyLocalStore(
+    const MaybeDocumentMap& changes,
+    const absl::optional<RemoteEvent>& maybe_remote_event) {
+  std::vector<ViewSnapshot> new_snapshots;
+  std::vector<LocalViewChanges> document_changes_in_all_views;
+
+  for (const auto& entry : query_views_by_query_) {
+    const auto& query_view = entry.second;
+    const View& view = query_view->view();
+    ViewDocumentChanges view_doc_changes = view.ComputeDocumentChanges(changes);
+    if (view_doc_changes.needs_refill()) {
+      // The query has a limit and some docs were removed/updated, so we need to
+      // re-run the query against the local store to make sure we didn't lose
+      // any good docs that had been past the limit.
+      DocumentMap docs = [local_store_ executeQuery:query_view->query()];
+      view_doc_changes =
+          view.ComputeDocumentChanges(docs.underlying_map(), view_doc_changes);
+    }
+
+    absl::optional<TargetChange> target_changes;
+    if (maybe_remote_event.has_value()) {
+      const RemoteEvent& remote_event = maybe_remote_event.value();
+      auto it = remote_event.target_changes().find(query_view->target_id());
+      if (it != remote_event.target_changes().end()) {
+        target_changes = it->second;
+      }
+    }
+    ViewChange view_change =
+        query_view->view().ApplyChanges(view_doc_changes, target_changes);
```

Comment:
Nit: there's a `view` local variable above, either use it here, or perhaps remove it?

Reply:
changed `view` to non-const to be able to reuse.